### PR TITLE
[Runtime] Extend Graph Runtime To Support Cuda Graph Launch

### PR DESCRIPTION
**Pull request by zhuochenKIDD**
We are currently using graph runtime to run some CTR models on NV-GPU,  for our in-house model (around 100 nodes in tvm json graph ) cuGraphLaunch can reduce 5% to 10% percent latency vs the original for-loop cuda kernel launch. 

So I wonder if the extension might benefits other workloads,  I haven't test other types of models. 

This is a POC, will supplement demos/docs

### CODE REVIEW

File: include/tvm/runtime/c_runtime_api.h [hunk 1]
```diff
@@ -481,6 +481,15 @@ TVM_DLL int TVMStreamFree(int device_type, int device_id, TVMStreamHandle stream
  */
 TVM_DLL int TVMSetStream(int device_type, int device_id, TVMStreamHandle handle);
 
+
+TVM_DLL int TVMStreamBeginCapture(int device_type, int device_id, TVMStreamHandle stream);
```
Comment (tqchen) [inline include/tvm/runtime/c_runtime_api.h:485]: Given cuGraph is specific to CUDA atm, let us not introduce it to the DeviceAPI for now, instead just use the cuda api in cugraph runtime

**Review by tqchen — CHANGES_REQUESTED** [review]

**Comment by tqchen** [comment]
cc @trevor-m @antinucleon @merrymercy please help to review

File: include/tvm/runtime/c_runtime_api.h (hunk 1 above)
Comment (zhuochenKIDD) [reply include/tvm/runtime/c_runtime_api.h]: thx, already removed

**Comment by tkonolige** [comment]
@zhuochenKIDD Do you have a guess as to why this is faster than the for-loop launch approach?

**Comment by zhuochenKIDD** [comment]
> @zhuochenKIDD Do you have a guess as to why this is faster than the for-loop launch approach?

@tkonolige not sure why it's faster，it's based on test，and depends on workloads. 
I guest for my model, it has many tiny kernels which is more kernel-launch bound, and cuda-graph might reduce kernel-launch overhead, I will do more profiling & analysis，do you have some suggestion?

**Comment by tqchen** [comment]
@zhuochenKIDD please checkout the lint error from the CI.

**Comment by comaniac** [comment]
In terms of the interface, can we use `GraphRuntimeFactory` to dispatch the extended runtime so that it would be more straightforward for users?

cc @FrozenGene @zhiics @icemelon9 @vinx13

**Comment by FrozenGene** [comment]
> In terms of the interface, can we use `GraphRuntimeFactory` to dispatch the extended runtime so that it would be more straightforward for users?
> 
> cc @FrozenGene @zhiics @icemelon9 @vinx13

If i don't understand it wrong, graph runtime we have done it before, but have not done for VM or other runtime.

**Comment by FrozenGene** [comment]
@zhuochenKIDD Besides requires CUDA 10, does cuda graph require special version of GPU hardware? like Tesla or Ampere?

**Comment by FrozenGene** [comment]
> > @zhuochenKIDD Do you have a guess as to why this is faster than the for-loop launch approach?
> 
> @tkonolige not sure why it's faster，it's based on test，and depends on workloads.
> I guest for my model, it has many tiny kernels which is more kernel-launch bound, and cuda-graph might reduce kernel-launch overhead, I will do more profiling & analysis，do you have some suggestion?

I think the answer is here: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-graphs

> This allows a graph to be defined once and then launched repeatedly. Separating out the definition of a graph from its execution enables a number of optimizations: first, CPU launch costs are reduced compared to streams, because much of the setup is done in advance; second, presenting the whole workflow to CUDA enables optimizations which might not be possible with the piecewise work submission mechanism of streams.

**Comment by FrozenGene** [comment]
> @zhuochenKIDD Besides requires CUDA 10, does cuda graph require special version of GPU hardware? like Tesla or Ampere?

I find the answer: https://docs.nvidia.com/cuda/cuda-samples/index.html#simple-cuda-graphs no special version of GPU hardware required. It is optimization at software level.

**Comment by comaniac** [comment]
> If i don't understand it wrong, graph runtime we have done it before, but have not done for VM or other runtime.

This PR is for graph runtime so it's not an issue for other runtimes? My concern is if we can reduce the API complexity from users. Now users need to use `tvm.contrib.cu_graph.cugraph_runtime.create(...)` to make use of this runtime, but this is not recommeneded. We should make all graph runtimes craeted from `graph_runtime.GraphModule`.

**Comment by FrozenGene** [comment]
> > If i don't understand it wrong, graph runtime we have done it before, but have not done for VM or other runtime.
> 
> This PR is for graph runtime so it's not an issue for other runtimes? My concern is if we can reduce the API complexity from users. Now users need to use `tvm.contrib.cu_graph.cugraph_runtime.create(...)` to make use of this runtime, but this is not recommeneded. We should make all graph runtimes craeted from `graph_runtime.GraphModule`.

Not it. I just meant our `GraphRuntimeFactory` has designed for this purpose you mentioned before. However, `GraphRuntimeFactory` is applied only for `GraphRuntime` now, not applied for VM or other runtime. But `GraphRuntimeFactory` could be used for the purpose you mentioned.

**Comment by zhuochenKIDD** [comment]
I opened a more clean PR, close this.
https://github.com/apache/tvm/pull/7616

**Pull request closed without merging**